### ISSUE TITLE
Use SASS placeholders for element definitions

### ISSUE THREAD

**Issue by nwpappas** [material-components/material-components-web#3357]
Using placehoders (`%class`) to hold the core styles for each element would allow design systems based on MDC to reduce their footprint, as well as be more tightly branded to itself.

Currently, MDC uses standard classes to hold the core styles and asks designers/developers to override styles through overload classes (e.g., `.my-custom-button`). This works, but has several drawbacks in a custom branded design system.

1. Two classes are required whenever an element is created, added complexity and potential for error.
    ```
    <button class="mdc-button my-customer-button"></button>
    ```
2. Unused MDC styles are still included.

In the case of item 2, in this _simplified_ example, imagine a custom design system that does not allow outlined buttons. Those definitions are still in the code and still have the potential to be used by designers/developers (who don't always listen to guidelines). The extra CSS also adds to file size.

## Proposal

Using SASS placeholders would allow a custom design system to brand itself fully, and tighten its implementation.

**_placeholders.scss**
```
%mdc-button {
  // ...
}
```

**mdc-button.scss**
```
.mdc-button {
  @extend %mdc-button;
}
```

**my-button.scss**
```
@import '@material/button/placeholders';

.my-button {
  @extend %mdc-button;
}
```

1. The original `.mdc-button` class is no not included in the CSS output.
2. The custom design system can a more easily remove unwanted aspects/variants (e.g., outlined buttons).
3) The custom design system can fully brand itself internally (`.acme-button`) without reference to mdc elements.

**Comment by nwpappas** [comment]
I realize button is rather oversimplified. Using the latest `.mdc-menu` updates might be better. Since it has mixin calls, class overrides, standard properties, and variable references.

Also, nesting class overrides might be considered problematic with placeholders. But if a design system developer is wanting to fully extend MDC (i.e., fully remove .mdc-* classes), that added complexity might just be the price to pay.

**Comment by kfranqueiro** [comment]
First of all, thanks for the feedback!

We avoided `@extend` for a couple of reasons:

* A placeholder can't be used across `@media` query scopes
* It can be hard to follow or to predict what it will do (though this is probably more applicable when used with classes rather than placeholders)

We experimented with the idea of organizing all styles into mixins for a few components (e.g. Button and Checkbox which delegate to internal mixins), but in practice this also runs into a few complications:

* MDC Web follows BEM, and thus attempts to use the least specificity necessary in its selectors. In some cases for components with child elements, this means we select _only_ on the child element in the baseline styles (e.g. just `.mdc-checkbox__checkmark` rather than `.mdc-checkbox .mdc-checkbox__checkmark`). In order to support this idea, we would need to expose mixins for these sub-elements for users to hook up classes to manually as well, which seems like a lot of complexity for users and a lot to properly document that could easily seem overwhelming to the reader.
* Many components have multiple variants, in which case there would need to be a "base" mixin and a mixin for each variant (base being separate to avoid repeating the same styles in each variant). Users would need to ensure they `@include` both. (I suppose this isn't much different than including both the base and variant `mdc-` classes, but it's just as much a potential point of user error.)

Moreover, we promote the pattern of applying the `mdc-` class for MDC Web styles and applying custom classes for customizations, in order to easily disambiguate which styles originate from our code, and which styles are external.

I do empathize with your point about unused styles being included (especially if a given design has no use for some variants). I'm wondering if this is something we could standardize a pattern for via global variables (or a global map) without getting too messy. I'll bring that particular point up with the team and we can log something more focused on that if we decide to pursue it.
